src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -195,4 +204,129 @@ private static void validateInputLines(String[] lines) {
+         "-Xshare:",
+         "-XX:SharedClassListFile=",
+         "-XX:SharedArchiveFile=",
+         "-XX:ArchiveClassesAtExit=",
+         "-XX:+UseSharedSpaces",
+         "-XX:+RequireSharedSpaces"};
+    private static boolean containsExcludedFlags(String testStr) {
+       for (String e : excludeFlags) {
+           if (testStr.contains(e)) {
+               return true;
+           }
+       }
+       return false;
+    }
+
+    /**
+    * called from jcmd VM.cds to dump static or dynamic shared archive
+    * @param isStatic true for dump static archive or false for dynnamic archive.
+    * @param fileName user input archive name, can be null.
+    */
+    private static void dumpSharedArchive(boolean isStatic, String fileName) throws Exception {
+        String currentPid = String.valueOf(ProcessHandle.current().pid());
+        String archiveFile =  fileName != null ? fileName :
+            "java_pid" + currentPid + (isStatic ? "_static.jsa" : "_dynamic.jsa");
+
+        // delete if archive file aready exists
+        File fileArchive = new File(archiveFile);
+        if (fileArchive.exists()) {
+            fileArchive.delete();
+        }
+        if (isStatic) {
+            String listFile = archiveFile + ".classlist";
+            File fileList = new File(listFile);
+            if (fileList.exists()) {
+                fileList.delete();
+            }
+            dumpClassList(listFile);
+            String jdkHome = System.getProperty("java.home");
+            String classPath = System.getProperty("java.class.path");
+            List<String> cmds = new ArrayList<String>();
+            cmds.add(jdkHome + File.separator + "bin" + File.separator + "java"); // java
+            cmds.add("-cp");
+            cmds.add(classPath);
+            cmds.add("-Xlog:cds,class+load");
+            cmds.add("-Xshare:dump");
+            cmds.add("-XX:SharedClassListFile=" + listFile);
+            cmds.add("-XX:SharedArchiveFile=" + archiveFile);
+
+            // All runtime args.
+            String[] vmArgs = VM.getRuntimeArguments();
+            if (vmArgs != null) {
+                for (String arg : vmArgs) {
+                    if (arg != null && !containsExcludedFlags(arg)) {
+                        cmds.add(arg);
+                    }
+                }
+            }
+
+            Process proc = Runtime.getRuntime().exec(cmds.toArray(new String[0]));
+
+            // Drain stdout to a file in a separate thread.
+            String stdOutFile = "java_pid" + proc.pid() + "_stdout.txt";
+            new Thread( ()-> {
+                    outputStdStream(proc.getInputStream(), stdOutFile, cmds);
+                }).start();
+
+            // Drain stderr to a file in a separate thread.
+            String stdErrFile = "java_pid" + proc.pid() + "_stderr.txt";
+            new Thread( ()-> {
+                    outputStdStream(proc.getErrorStream(), stdErrFile, cmds);
+                }).start();
+
```

Comment:
I think the above code can be refactored to avoid duplication. Something like:

```
String stdOutFile = drainOutput(proc.getInputStream(), "stdout");
String stdErrFile = drainOutput(proc.getErrorStream(), "stderr");
```

and move the common code into drainOutput().